inductiva/templating/renderers.py

```diff
@@ -0,0 +1,120 @@
+"""Base and specific renderers packages to manage template files."""
+from abc import abstractmethod
+from typing import Iterable, Union, IO
+import pathlib
+import shutil
+import os
+
+import jinja2
+
+from inductiva import types, utils
+
+
+class BaseRenderer:
+    """Abstract base class for template renderers."""
+
+    TEMPLATE_EXTENSION = None
+
+    @abstractmethod
+    def render_dir(self, target_dir: types.Path, source_dir: types.Path,
+                   **render_args):
+        """Render a template to a file."""
+        pass
+
+    @abstractmethod
+    def render_file(self, target_file: types.Path,
+                    source_file: Union[types.Path, IO], **render_args):
+        """Render a file."""
+        pass
+
+    @classmethod
+    def is_template(cls, file: types.Path) -> bool:
+        """Check if the given file is of template type."""
+        if isinstance(file, pathlib.Path):
+            return file.suffix == cls.TEMPLATE_EXTENSION
+        return file.endswith(cls.TEMPLATE_EXTENSION)
+
+    @classmethod
+    def strip_extension(cls, file: types.Path) -> str:
+        """Strip the template extension if the given file is a template."""
+        if not cls.is_template(file):
+            return file
+        if isinstance(file, pathlib.Path):
+            return file.with_suffix("")
+        return os.path.splitext(file)[0]
+
+    @classmethod
+    def get_template_files(cls, source_dir: types.Path) -> Iterable[types.Path]:
+        """Return a list of template files in the given source directory."""
+
+        return utils.files.map_dir_files(source_dir, cls.TEMPLATE_EXTENSION)
+
+
+class JinjaRenderer(BaseRenderer):
+    """Jinja2 renderer for .jinja templates."""
+
+    TEMPLATE_EXTENSION = ".jinja"
+
+    @classmethod
+    def render_dir(cls, source_dir: types.Path, target_dir: types.Path,
+                   **render_args):
+        """Render a directory from a template directory.
+
+        Render the jinja template files within a given source_dir into a 
+        target_dir. These files are rendered with the given render_args and
+        the rendered files keep the same name without the .jinja extension.
+
+        Note, all non-template files are also copied to the target directory.
+
+        Args:
+            source_dir (types.Path): Path to the source directory.
+            target_dir (types.Path): Path to the target directory.
+            render_args: Arguments to render the template files.
+        """
+        target_dir = pathlib.Path(target_dir)
+
+        shutil.copytree(source_dir,
+                        target_dir,
+                        dirs_exist_ok=True,
+                        symlinks=True)
+
+        template_dir_struct = cls.get_template_files(target_dir)
+
+        # Render the template files from within the target directory.
+        environment = jinja2.Environment(
+            loader=jinja2.FileSystemLoader(target_dir),
+            undefined=jinja2.StrictUndefined)
```

Comment:
this should be unique to the instance. the shouldn't be any need to instantiate a new environment every time a file is rendered. Look at how it was done in https://github.com/inductiva/inductiva/pull/1160/files. It was pretty stable and clean.

Reply:
This is done, because we can template different dirs now, with the `source_dir` being passed. In the PR you link, we were stucking the user to have to instantiate a new object of Template type, every time we would want to change the template_dir. So this brings the best of both worlds and allows the user to have a single base template dir with all his jinja files.

Reply:
The `TemplateEngine` class is instantiated with a template dir which is supposed to be the source of all templates. It makes sense to point the environment to that dir.